cura/Settings/MachineManager.py

```diff
@@ -1343,25 +1270,17 @@ def updateMaterialWithVariant(self, position: Optional[str]) -> None:
             if extruder.variant.getId() != empty_variant_container.getId():
                 current_nozzle_name = extruder.variant.getMetaDataEntry("name")
 
-            material_diameter = extruder.getCompatibleMaterialDiameter()
-            candidate_materials = self._material_manager.getAvailableMaterials(
-                self._global_container_stack.definition,
-                current_nozzle_name,
-                buildplate_name,
-                material_diameter)
-
-            if not candidate_materials:
-                self._setMaterial(position_item, container_node = None)
-                continue
-
-            if current_material_base_name in candidate_materials:
+            # If we can keep the current material after the switch, try to do so.
+            nozzle_node = ContainerTree.getInstance().machines[self._global_container_stack.definition.getId()].variants[current_nozzle_name]
+            candidate_materials = nozzle_node.materials
+            if current_material_base_name in candidate_materials:  # The current material is also available after the switch. Retain it.
                 new_material = candidate_materials[current_material_base_name]
                 self._setMaterial(position_item, new_material)
-                continue
-
-            # The current material is not available, find the preferred one
-            material_node = self._material_manager.getDefaultMaterial(self._global_container_stack, position_item, current_nozzle_name)
-            if material_node is not None:
+            else:
+                # The current material is not available, find the preferred one.
+                material_diameter = self._global_container_stack.extruders[position].getCompatibleMaterialDiameter()
+                approximate_material_diameter = round(material_diameter)
```

Comment:
Same as above, you can directly use `getApproximateMaterialDiameter()`

Reply:
Done: 963f5e6dd73d780ce19f5cc02a0d22f22ce696aa